storing addresses ranges which kernel accesses within every syscall

Merna177 opened the pull request:
The first step in Double fetch problem is to get intersection between memory ranges which kernel accesses within system call so i can know where the problem occurs  
Here, i add struct to store starting address and how many bytes which kernel accesses during every system call, so from here i can know the range of addresses.

Commit:
storing addresses ranges which kernel accesses within every syscall

ramosian-glider commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+};
+
+
+void addAddress(void* addr,unsigned long len);
```
I suggest you take a look at https://www.kernel.org/doc/html/v4.10/process/coding-style.html
Your style is ok for any other language (or even another C project), but it is inconsistent with the rest of the kernel.
In particular, look at the "Naming" section.
E.g. this function should be called `add_address`, the next one - start_system_call, etc.

It's important that we fix this all now, before you write too much code, because the kernel maintainers will make us change the style anyway :)

Merna177 replied:
Okay i will fix the naming and follow the naming conventions.
Thanks

ramosian-glider commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+#ifdef CONFIG_DF_DETECTION
+
+struct address{
+    void *startaddress;
```
Nit: I'd probably suggest `start_address`, `start_addr`, `start`, or `addr` for this one. Your call.

ramosian-glider commented on include/linux/sched.h:
```diff
@@ -34,6 +34,7 @@
 #include <linux/rseq.h>
 #include <linux/seqlock.h>
 #include <linux/kcsan.h>
+#include "linux/double-fetch-detection.h"
```
Compared to other LInux headers, this name is a bit too long. Something like "df-detection.h" should be enough.

Again, please only consider these naming conventions when writing patches for the Linux kernel. It is perfectly fine in other projects :)

melver replied:
I think they don't like abbreviations too much, because what is "df"? Same for config var.

Maybe at the end of all of it a nicer name will emerge, but I don't have a suggestion right now.

ramosian-glider replied:
Ok, I don't have any strong preference, leaving this up to Merna. Maybe we'll come up with a good name later.

ramosian-glider commented on init/main.c:
```diff
@@ -262,11 +262,10 @@ early_param("loglevel", loglevel);
 #ifdef CONFIG_BLK_DEV_INITRD
 static void * __init get_boot_config_from_initrd(u32 *_size, u32 *_csum)
-{
+{  
```
When working with Git, check regularly that you didn't change code that you weren't going to change.
E.g. the whole diff to this file could be reverted, as it is unrelated to your changes.

Merna177 replied:
yeah right i am sorry for that.

ramosian-glider commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+
+void addAddress(void* addr,unsigned long len);
+void startSystemCall(void);
+void endSysCall(void);
```
For paired names use consistent naming, e.g. `endSystemCall` instead of `endSysCall` (or, in our case, `end_system_call`)

ramosian-glider commented on log.txt:
```diff
@@ -0,0 +1,2732 @@
+  DESCEND  objtool
```
You probably don't want to keep this log file in Git.
It can be removed from the index with `git rm --cached log.txt`

ramosian-glider commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+}
+
+void startSystemCall(){
+    current->addresses= (struct address*)kmalloc_array(100,sizeof(struct address),GFP_KERNEL);
```
Hm, actually, as we're already allocating the memory dynamically, let's start with a small number of records (say, 16), and then reallocate the array if we need more.
You'll need to store the array size in `current` as well.

Merna177 replied:
okay i did that and every time it needs more memory i reallocate and make the size its double like(16,32,64..)
but can i make an upper bound for the size INT_MAX?

ramosian-glider replied:
Hm, interesting. Can you call dump_stack() to see in which cases this happens?

I didn't expect the number of accesses to be bigger than IOV_MAX+10 (for `readv()/writev()`, which probably have the most number of pointers passed to them), but maybe I am missing something. E.g. the kernel may be splitting the buffer passed to `write()` in fixed-size chunks (no idea if this is actually happening).

ramosian-glider replied:
Sorry, I misunderstood Merna's previous comment.
We've figured out in an offline chat that there's no problem with doubling the buffer size. I also suggested to set a decent upper limit (say, 1024 records) and warn if the syscall requests more (e.g. by calling `WARN_ON()`)

ramosian-glider commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+#include <linux/audit.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+void addAddress(void* addr, unsigned long leng){
```
Nit: `len` is a more common name than `leng`

ramosian-glider commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+#include <linux/sched.h>
+#include <linux/slab.h>
+void addAddress(void* addr, unsigned long leng){
+           if(current->noRead==100 || current->noRead==-1)
```
Nit: the name `noRead` has a connotation with something being not read. How about `num_read`?

ramosian-glider commented on include/linux/sched.h:
```diff
@@ -1314,7 +1315,10 @@ struct task_struct {
 	struct callback_head		mce_kill_me;
 #endif
-
+    //needed in double fetch detection buffer to store address ranges 
```
Please put this part insude `#ifdef CONFIG_DF_DETECTION`

ramosian-glider commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+void addAddress(void* addr, unsigned long leng){
+           if(current->noRead==100 || current->noRead==-1)
+             return;
+           ((current->addresses)+current->noRead)->startaddress = addr;
```
It's more convenient to use indexing, i.e.:
`current->addresses[current->noRead]->startaddress`

ramosian-glider reviewed (commented):
Nice! Please go on.

melver commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+#include <linux/sched.h>
+#include <linux/slab.h>
+void addAddress(void* addr, unsigned long leng){
+           if(current->noRead==100 || current->noRead==-1)
```
What do the constants mean? They should probably be named constants defined somewhere else.

Merna177 replied:
i think after reallocation if i need more memory i can delete constant"100".
also i can replace the checking of negative one to check on the pointer if it still not pointing to any address that means no system call need userspace address yet.

melver commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
+#ifdef CONFIG_DF_DETECTION
+
+struct address{
```
This should be prefixed with something to denote its "namespace".

E.g. "double_fetch_detection_address" although that's way too long ...

Merna177 replied:
Can it be user_space_address or user_address_range? as it refers to addresses from userspace?

melver replied:
I think it's still too general. I highly doubt anybody else will reuse this, because all you have here is a pointer+size, but otherwise nothing specific to addresses from userspace.

The usual convention is that for public identifiers across headers in include/linux is to prefix or somehow add the subsystem into each name (unless it's a core kernel primitive).

df_address_range might work, but I still don't like 'df' because many things could be 'df'.

melver commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+
+struct address{
+    void *startaddress;
+    unsigned long len;
```
why not 'size_t'?

Merna177 replied:
as far as i know size_t is unsigned long ?  is defined in include/linux/types.h as unsigned long. 
i wrote it unsigned long just like the datatype of the parameter (n) in  copy_from_user function

melver replied:
If you want consistency with the copy..user interface, 'unsigned long' is fine. But I also do not see why we can't just use size_t here. I'd probably use size_t, just because it's a little more precise and there should be no harm in doing so.

melver commented on include/linux/double-fetch-detection.h:
```diff
@@ -0,0 +1,17 @@
+};
+
+
+void addAddress(void* addr,unsigned long len);
```
Should probably be 'const void*'

melver commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+#include "linux/double-fetch-detection.h"
```
To fix-up your new files' formatting just run clang-format on them. The kernel has a .clang-format style file that's pretty good to get started.

ramosian-glider replied:
Some context: https://clang.llvm.org/docs/ClangFormat.html

You probably have clang-format in your system's package manager.

melver commented on log.txt:
```diff
@@ -0,0 +1,2732 @@
+  CC      lib/usercopy.o
+  CC      drivers/acpi/acpica/rscreate.o
+lib/usercopy.c: In function ‘_copy_from_user’:
+lib/usercopy.c:17:14: warning: passing argument 1 of ‘addAddress’ discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
```
Fix the function signature to get rid of the warnings.

melver commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+
+void startSystemCall(){
+    current->addresses= (struct address*)kmalloc_array(100,sizeof(struct address),GFP_KERNEL);
+    if(!current->addresses)
```
```
current->noRead = current->addresses ? 0 : -1;
```
is shorter.

melver commented on lib/double-fetch-detection.c:
```diff
@@ -0,0 +1,34 @@
+
+}
+void endSysCall(){
+    print();
```
Shouldn't this warn if you reach some threshold of read accesses? Right now it seems it always prints something if noRead>0.

Will this still change?

Merna177 replied:
Print function i used only for testing or printing anything i need during debugging.

Commit:
revert the changes in init/main.c

Commit:
revert changes in fork.c

Commit:
change naming style for the functions and the variables, replace allocating a large memory in the beginning with reallocating memory when needed, replace the constants and delete unneeded files

Merna177 commented:
> Nit: constant names should be all uppercase.

Merna177 commented:
Thank you so much for the comments i learnt from it alot, they were so useful, i tried to address all the comments and fixed it so i pushed all the changes ,i hope that i am not missing anything from the comments i reviewed the code styling and all other comments ,delete any unneeded code and fixed the bug KASAN_DOUBLE_FREE.

Commit:
fix code styling

Merna177 commented:
Thank you so much for the comments i learnt from it alot, they were so useful, i tried to address all the comments and fixed it so i pushed all the changes ,i hope that i am not missing anything from the comments i reviewed the code styling and all other comments and delete any unneeded code.

Commit:
fixed the indentation of include/linux/sched.h file

Commit:
Adding intersection function to detect the overlapping region in address ranges and also modifed reallocate_extra_memory to check if we exceed the limits of the size

Commit:
Modifed reallocate_extra_memory function

Commit:
fixed reallocate_extra_memory function

Merna177 commented:
i did changes to reallocate_extra_memory function to make it only check if it valid to allocate extra memory or not

Commit:
Modified intersection function to detect any intersection while adding new address ranges on the fly

Commit:
Adding print for testing purpose

ramosian-glider commented on lib/Kconfig.debug:
```diff
@@ -34,6 +34,10 @@ config PRINTK_CALLER
 	  Since toggling after boot makes the code racy, currently there is
 	  no option to enable/disable at the kernel command line parameter or
 	  sysfs interface.
+	  
```
Please remove whitespaces from lines 37 and 38, kconfig is complaining: `lib/Kconfig.debug:40:warning: leading whitespace ignored`

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,116 @@
+	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
+	current->df_index = 0;
+}
+void end_system_call()
```
According to the kernel style guide, this should be written as `void end_system_call(void)`.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,116 @@
+	}
+}
+
+void start_system_call()
```
I think it makes sense to pass the syscall number to this function, otherwise we actually don't have any information about which syscall we are looking at.

ramosian-glider replied:
This might require calling start_system_call() from syscall_enter_from_user_mode(), where we know the syscall number.

Merna177 replied:
yeah that what i did

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,116 @@
+	if (current->pairs != NULL) {
+		if (current->df_index) {
+			print_pairs();
+			dump_stack();
```
Turns out dump_stack() is useless: the stack is not specific to the syscall or call site:
```
Call Trace:
 dump_stack+0x57/0x70 ??:?
 end_system_call+0x9c/0xb0 ??:?
 syscall_exit_to_user_mode+0x2b/0x160 ??:?
 entry_SYSCALL_64_after_hwframe+0x44/0xa9 ??:?
```

What we'll need to report is the syscall number (or name) and/or the callers of `add_address()`.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,116 @@
+		if (current->df_index) {
+			print_pairs();
+			dump_stack();
+			pr_err("BUG: Intersection Detected \n ");
```
This line should be the first in the report: syzkaller will cut everything above.

Commit:
saving the caller of copy_from_user function

Commit:
adding system call number

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -7,7 +7,7 @@
 #include <linux/stacktrace.h>
 
-void add_address(const void *addr, unsigned long len)
+void add_address(const void *addr, unsigned long len,unsigned int caller)
```
Nit: space after comma.

ramosian-glider replied:
Also, `caller` must be unsigned long (see definition of `_RET_IP_`: https://elixir.bootlin.com/linux/latest/source/include/linux/kernel.h#L170)
If you pass it as unsigned int, the top bytes will be truncated, which may result in an incorrect address being stored/printed.

ramosian-glider replied:
And you need to actually store `caller` in `current->addresses`

Merna177 replied:
okay i will change it to unsigned long sorry for that.
i already saved the caller in addresses struct

ramosian-glider commented on lib/usercopy.c:
```diff
@@ -14,7 +14,7 @@ unsigned long _copy_from_user(void *to, const void __user *from, unsigned long n
 		instrument_copy_from_user(to, from, n);
 		res = raw_copy_from_user(to, from, n);
-		add_address(from, n);
+		add_address(from, n,_RET_IP_);
```
Nit: space after comma.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -61,7 +61,7 @@ void end_system_call()
 	}
 }
-void print_pairs()
+void print_pairs(void)
```
Need to modify `print_pairs()` to print call sites for intersecting ranges.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -25,13 +25,16 @@ void add_address(const void *addr, unsigned long len,unsigned int caller)
+void start_system_call(long syscall)
 {
+	current->syscall_num = syscall;
+	pr_err("syscall %ld",syscall);
```
Nit: we'd better print the syscall number when reporting the error.
Also, printing `sys_call_table[current->df_syscall]` would be more readable than just a raw number.

Commit:
adding error report

melver commented on arch/x86/entry/common.c:
```diff
@@ -52,6 +52,7 @@ __visible noinstr void do_syscall_64(unsigned long nr, struct pt_regs *regs)
 		regs->ax = x32_sys_call_table[nr](regs);
 #endif
 	}
+
```
revert

ramosian-glider replied:
+1

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -48,9 +47,7 @@ void end_system_call(void)
-			print_pairs();
-			dump_stack();
-			pr_err("BUG: Intersection Detected \n ");
+			report();
```
Nit: if your if-body consists of only one line, you can omit the braces, i.e.:
```
if (current->df_index)
	report();
```

melver commented on include/linux/sched.h:
```diff
@@ -1323,13 +1333,15 @@ struct task_struct {
 	 *
 	 * Do not put anything below here!
 	 */
+	
```
revert newline chances

ramosian-glider replied:
+1

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -64,17 +61,27 @@ void end_system_call(void)
+	pr_err("BUG: Intersection Detected \n ");
+	pr_err("==================================================================\n");
+	if (panic_on_warn) {
+		pr_err("System Call Number: %ld\n", current->syscall_num);
```
I still think printing the symbol from sys_call_table[] is worthwhile here.

Merna177 replied:
yeah i know i will use sys_call_table instead of just printing systemcall number, but i just push the updates after making sure that this report function is printing in syzkaller crash report. to make everything up-to-date.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -64,17 +61,27 @@ void end_system_call(void)
+			       current->pairs[i].second->len,
+			       current->pairs[i].second->caller);
+		}
+		pr_err("==================================================================\n");
```
As discussed today, try panicking here, so that we get syzkaller to work.

Merna177 replied:
is it panic_on_warn not enough ? i know that we discussed to add panic to the function report but i thought that because panic_on_warn wasn't work but now i fix what i did wrong and it printing all the data we need in syzkaller report

ramosian-glider replied:
Yes, panic_on_warn is also enough, but it should only be checked before calling panic().
The report must be printed unconditionally, i.e. if you start your kernel without panic_on_warn, you should see a bunch of reports in the log, but if you set panic_on_warn then you'll crash after the first one.

Commit:
remove new lines

Commit:
using sys_call_table instead of systemcall number to be more readable

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -67,7 +67,8 @@ void report(void)
 	pr_err("==================================================================\n");
 	if (panic_on_warn) {
-		pr_err("System Call Number: %ld\n", current->syscall_num);
+		pr_err("System Call: %ps\n",
```
Cool!

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -67,7 +67,8 @@ void report(void)
 	pr_err("BUG: Intersection Detected \n ");
 	pr_err("==================================================================\n");
 	if (panic_on_warn) {
```
We actually need to always print the report (regardless of whether we want panic_on_warn or not), but then call panic() if panic_on_warn is set.

Merna177 replied:
but when i tried to print report without panic_on_warn non of the prints appear on syzkaller report , only when i used panic_on_warn all the data we wanted appears in syzkaller report

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -67,7 +67,8 @@ void report(void)
 	pr_err("BUG: Intersection Detected \n ");
```
Note that syzkaller treats whatever you write after "BUG:" as a report header, and it clusters together reports with similar headers.
So we probably want to at least print the syscall function name on the same line - this way we'll get different different syscalls bucketed separately.
In the future we'll probably want to go even further and add the call sites to the header to separate bugs from each other.

Commit:
changes in report function to make a report for every systemcall

Commit:
store the address ranges from all functions that copy from user space

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,133 @@
+	pr_err("==================================================================\n");
+	if (panic_on_warn) {
+		panic("panic_on_warn set. \n");
+		panic_on_warn = 0;
```
This line has no effect, because the kernel is dead at this point already. Please move it before panic().

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,133 @@
+}
+
+// it returns 0 when it fails to re allocate memory
+int reallocate_extra_memory(int sz, int mx_size)
```
I think you don't need this function anymore.

Merna177 replied:
in syzkaller detected that this warning happened 4 times

ramosian-glider replied:
True; what I was trying to say is that the function doesn't really match its name anymore, it just checks the size instead of actually reallocating memory.
You could either perform the check in-place instead of calling this function (it's basically equivalent to `!WARN_ON(sz, mx_size)`, or pull krealloc into this function as well, so that it encapsulates memory reallocation.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,133 @@
+
+void start_system_call(long syscall)
+{
+	if (syscall == 54 || syscall == 165 || syscall == 16)
```
I think we'd better keep this logic in report(), or maybe introduce a special function that tells us whether to print the report or not. In the future the filtering rules will become more complex and will probably depend on other data that's not available to start_system_call().

Commit:
adding stack depot to save the stack after every call to add_address function

Commit:
change condition in check_valid_detection function

Commit:
change condition in check_valid_detection function

Commit:
add interested function calls to the header of the report

Commit:
add interested functions to the header of the report

Commit:
changes in the report and add add_address function in all copy from userspace functions

Commit:
called add_address from strncpy function

Commit:
fixed UAF Bug

Commit:
ignore copy_from_user_nmi

Commit:
neglect case if 2 fetches happened in perf_copy_attr

Commit:
add debugfs_create_file

Commit:
add comments and change value of max len

Commit:
adding df_enable and df_disable options

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -159,7 +158,8 @@ int filter_stack(const unsigned long stack_entries[], int num_entries)
 		    strnstr(buf, "copyin", len) ||
 		    strnstr(buf, "strncpy_from_user", len) ||
-		    strnstr(buf, "get_user", len))
+		    strnstr(buf, "get_user", len) ||
```
By the way, do you actually see occurrences get_user() in the stacks? I've never seen them - probably because that's a macro, not a function.

Merna177 replied:
some Intersection BUGS are printed during rebooting in syzkaller contain get_user in its stack system calls were (ioctl, execve), but i am not pretty sure is there any crashes detected after rebooting have get_user in its stack i can know that after disabling DF_DETECTION during rebooting.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -220,4 +220,18 @@ void detect_intersection(void)
+	return nonseekable_open(inode, filep);
+}
+static const struct file_operations df_fops = {
+    .open = df_open,
```
You don't need df_open, as it's not doing anything:
`.open = nonseekable_open`

Merna177 replied:
yeah you are right, i will change it.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -225,8 +226,24 @@ static int df_open(struct inode *inode, struct file *filep)
 {
 	return nonseekable_open(inode, filep);
 }
+static long df_ioctl(struct file *filep, unsigned int cmd, unsigned long arg)
```
Please try making `cmd` an unsigned long and remove `arg`

ramosian-glider replied:
Or, if `arg` is required, rename it to something like `unused`

Merna177 replied:
i understand i think we need the third parameter and the cmd to be unsigned int as the new way of ioctl which is unlocked_ioctl. and compat_ioctl are required 3 arguments like these:
 long (*unlocked_ioctl) (struct file *filp, unsigned int cmd, 
                            unsigned long arg);
    long (*compat_ioctl) (struct file *filp, unsigned int cmd, 
                          unsigned long arg);

Commit:
make df_enable false by default

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -44,27 +44,29 @@ void add_address(const void *addr, size_t len, unsigned long caller)
-	    DF_INIT_SIZE, sizeof(struct df_pair), GFP_KERNEL);
-	current->df_size = current->pairs ? DF_INIT_SIZE : 0;
-	current->df_index = 0;
+	if (current->df_enable) {
```
Prefer early return to avoid unnecessary code indentation:
```
if (current->df_enable)
	return;
current->syscall_num = syscall;
...
```

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -44,27 +44,29 @@ void add_address(const void *addr, size_t len, unsigned long caller)
 void end_system_call(void)
 {
-	if (current->pairs != NULL) {
+	if (current->df_enable && current->pairs != NULL) {
```
Checking for the same condition twice often flags that the code can be simplified.
Early return is much more preferable here.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -172,7 +174,7 @@ int filter_stack(const unsigned long stack_entries[], int num_entries)
 {
 	if (current->syscall_num == 54 || current->syscall_num == 165 ||
-	    current->syscall_num == 47)
+	    current->syscall_num == 55)
```
The comment above needs to be updated.

Commit:
ignore kcov_ioctl multi-read crashes

Commit:
ignore kcov_ioctl overlap addresses

Commit:
Merge branch 'storing-addresses' of https://github.com/Merna177/linux into storing-addresses

Commit:
make an early return instead unnecessary if-condition

ramosian-glider commented on lib/Kconfig.debug:
```diff
@@ -34,6 +34,9 @@ config PRINTK_CALLER
 	  Since toggling after boot makes the code racy, currently there is
 	  no option to enable/disable at the kernel command line parameter or
 	  sysfs interface.
+	    
```
Kconfig complains about the whitespaces in this line, please remove them.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,259 @@
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+#include <linux/sys.h>
+#include <linux/types.h>
```
Nit: please insert an empty line after this one.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,259 @@
+		detect_intersection();
+		current->num_read++;
+	}
+}
```
Nit: please leave a blank line between functions.

Commit:
changes in the style

Commit:
Moving add address after copying from userspace

Merna177 commented:
As we need to move add_address after copying process is done so i removed the add_address call from  include/linux/uaccess.h , because the function return raw_copy_from user function, and add the call of add_address to raw_copy_from_user.
In case we dont want to call add_address from raw_copy_function we can do something like this in copy from user functions in include/linux/uaccess.h : 
```
 res = raw_copy_from_user(to, from, n);
add_address(from, n, _RET_IP_);
return res; 
```
but by this we will change an exist function (i dont know if it is good idea or not)

Commit:
remove calling add_address from usercopy.c

Commit:
add some randomization to overlapped region

Commit:
fixed a bug in the randomization

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -50,6 +50,14 @@ void add_address(const void *addr, size_t len, unsigned long caller)
 	}
 }
 
+void add_randomization(const void *addr, size_t len)
```
As Dmitry pointed out, we shouldn't be doing this, because `addr` is a userspace address.
Because we are calling add_address() after the data has been copied from the userspace, it's already late to change it (the changes would be visible to the kernel only after the third fetch).

I think you need to pass the kernel memory pointer to add_address as well, and add_randomization should be changing the kernel buffer.

ramosian-glider commented on kernel/entry/common.c:
```diff
@@ -23,9 +24,8 @@ static __always_inline void enter_from_user_mode(struct pt_regs *regs)
 	lockdep_hardirqs_off(CALLER_ADDR0);
 
-	CT_WARN_ON(ct_state() != CONTEXT_USER);
+        CT_WARN_ON(ct_state() != CONTEXT_USER);
```
Please revert this part, it is unrelated to your patch.

ramosian-glider replied:
This line still doesn't match the upstream source.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,286 @@
+	    strnstr(buf_caller, "kcov_ioctl", size_caller))
+		return;
+	if (current->num_read >= current->sz &&
+	    !WARN_ON(current->sz > DF_MAX_RECORDS)) {
```
These warnings are causing syzkaller crashes and are not really actionable. Let's remove them and just skip the remaining addresses if we hit the limit.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,286 @@
+				  current->addresses[current->num_read]))
+			continue;
+		if (current->df_index >= current->df_size &&
+		    !WARN_ON(current->df_size >
```
Same here.

Commit:
remove exceed size limit warning

Commit:
modify kernel address instead

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -10,7 +10,8 @@
 #include <linux/types.h>
 
-void add_address(const void *addr, size_t len, unsigned long caller)
+void add_address(const void *addr, size_t len, unsigned long caller,
```
You also need to update the function prototype in include/linux/df-detection.h, without it the kernel doesn't compile.

Commit:
changes in add_address

Commit:
fix calling add_address

Commit:
remove calling add_address

Commit:
call add_address after copying data

Commit:
use random data instead of adding one to every byte

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,282 @@
+			    strnstr(second_buf, "perf_copy_attr", second_len))
+				continue;
+			pr_err("BUG: multi-read in %ps  "
+			       "between %ps and %ps\n ",
```
As Dmitry noted, let's try to print only first_frame in the header without the syscall name and second_frame.
This will help to cluster similar reports together more efficiently.

Merna177 replied:
so we need 2 frames from the stack of each address range after filtering copy functions?
so by this in the title we will have 4 functions name

Merna177 replied:
and also isn't syscall name in the header make it easier to link multi read crashes with fault injection crashes

ramosian-glider replied:
No, what I meant is just make the above `pr_err("BUG: multi-read in %ps  ", first_frame);`

But, fair enough, preserving the syscall name may help us better link multi-read reports with crashes. Feel free to just omit second_frame then, something along the lines of: `"BUG: multi-read in %ps, syscall %ps\n "`

Merna177 replied:
yeah okay sorry i misunderstood, okay i will remove only second_frame and i will keep syscall name if we observe that it won't help us linking between multi read and fault injection, i can easily remove it

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,282 @@
+		    current->addresses[current->pairs[i].first].caller,
+		    current->addresses[current->pairs[i].second].start_address,
+		    current->addresses[current->pairs[i].second].len,
+		    current->addresses[current->pairs[i].second].caller);
```
Please add a call to dump_stack_print_info() at the end of the report (after this line).
It will print some extra information about the CPU and process, e.g.:
```
[    2.226036] CPU: 0 PID: 95 Comm: kunit_try_catch Tainted: G        W         5.9.0-rc4-00178-gf5a65530e7de-dirty #1176
[    2.226174] Hardware name: linux,dummy-virt (DT)
```

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,282 @@
+		len = scnprintf(buf, sizeof(buf), "%ps",
+				(void *)stack_entries[indx]);
+
+		if (strnstr(buf, "copy_from_user", len) ||
```
Dmitry says this filtering can be moved to pkg/report in syzkaller, that is probably also worth looking into.

Commit:
add extra information in the report

Commit:
add check in_task

Merna177 committed:
changes in naming and remove some checks

ramosian-glider commented on arch/arm64/include/asm/uaccess.h:
```diff
@@ -23,6 +23,7 @@
 #include <asm/ptrace.h>
 #include <asm/memory.h>
 #include <asm/extable.h>
+#include "linux/df-detection.h"
```
Please move this line above, next to <linux/bitops.h>
Also, change ```""``` to ```<>```

ramosian-glider commented on arch/x86/include/asm/uaccess.h:
```diff
@@ -11,6 +11,7 @@
 #include <asm/page.h>
 #include <asm/smap.h>
 #include <asm/extable.h>
+#include "linux/df-detection.h"
```
Please move above, after linux/compiler.h and use angle brackets.

Overall, it is handy to sort the #includes alphabetically (see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes for an example).
Linux kernel doesn't have strict policies regarding this, but we try to avoid increasing the entropy by keeping newly added headers sorted.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef __LINUX_DOUBLE_FETCH_DETECTION_H__
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
```
Please add a blank line after this one.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#define MAX_LEN 1 << 20
+#define STACK_DEPTH 64
+#define DFETCH_ENABLE			_IO('c', 254)
+#define DFETCH_DISABLE			_IO('c', 255)
```
Please indent _IO on these two consequent lines similarly.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef __LINUX_DOUBLE_FETCH_DETECTION_H__
```
Please make this header look like existing ones (you can check include/linux/kcsan.h for an example).
First, add the SPDX license header at the top (`/* SPDX-License-Identifier: GPL-2.0 */`)
Second, mind the vertical spacing, add blank lines between logically separate parts.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef __LINUX_DOUBLE_FETCH_DETECTION_H__
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
+#ifdef CONFIG_DF_DETECTION
```
Please add a blank line here as well.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
+#ifdef CONFIG_DF_DETECTION
+#include <linux/types.h>
+#include <linux/stackdepot.h>
```
Blank line after the #include list, please.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#ifndef __LINUX_DOUBLE_FETCH_DETECTION_H__
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
+#ifdef CONFIG_DF_DETECTION
+#include <linux/types.h>
```
Please keep newly added #include directives sorted alphabetically.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#define STACK_DEPTH 64
+#define DFETCH_ENABLE			_IO('c', 254)
+#define DFETCH_DISABLE			_IO('c', 255)
+#define BYTE_MAX 256
```
Please add a blank line to separate #defines from struct definitions.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+#define DFETCH_ENABLE			_IO('c', 254)
+#define DFETCH_DISABLE			_IO('c', 255)
+#define BYTE_MAX 256
+struct df_address_range{
```
Please add a space between the struct name and the opening brace here and below.
Consider running clang-format to fix such nits altogether.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+void detect_intersection(void * kernel_addr);
+int is_intersect(struct df_address_range a, struct df_address_range b, void * kernel_addr);
+#endif
+#endif
```
Please add a comment after this #endif:

`#endif  /* __LINUX_DOUBLE_FETCH_DETECTION_H__ */`

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,35 @@
+bool check_valid_detection(void);
+void detect_intersection(void * kernel_addr);
+int is_intersect(struct df_address_range a, struct df_address_range b, void * kernel_addr);
+#endif
```
Please add a blank line before and after this #endif, also a comment after it: `#endif  /* CONFIG_DF_DETECTION */`

Overall, it's a good idea to mark the #endif with a comment when it is too far from the corresponding #ifdef (or if there are other #ifdef/#endif directives between them)

ramosian-glider commented on include/linux/sched.h:
```diff
@@ -34,6 +34,7 @@
 #include <linux/rseq.h>
 #include <linux/seqlock.h>
 #include <linux/kcsan.h>
+#include "linux/df-detection.h"
```
Please put this header where applicable.
Note that we're using angle brackets for the headers residing in include/linux.

ramosian-glider commented on include/linux/sched.h:
```diff
@@ -1315,6 +1316,17 @@ struct task_struct {
 
+ //needed in double fetch detection buffer to store address ranges 
+#ifdef CONFIG_DF_DETECTION
+        struct df_address_range *addresses;
```
Please fix indentation of the struct fields (only use tabs to indent struct fields in the kernel)

ramosian-glider commented on kernel/entry/common.c:
```diff
@@ -4,6 +4,7 @@
 #include <linux/entry-common.h>
 #include <linux/livepatch.h>
 #include <linux/audit.h>
+#include "linux/df-detection.h"
```
Please fix this one as well.

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -74,18 +74,24 @@ static inline void force_uaccess_end(mm_segment_t oldfs)
 	check_object_size(to, n, false);
-	return raw_copy_from_user(to, from, n);
+	res = raw_copy_from_user(to, from, n);
+	add_address(from, n, _RET_IP_, to);
```
For copy_from_user() we wanted to only store the ranges that were actually read/written.
Usercopy functions return the number of bytes that was _not_ read or written, so you'll need to adjust the `len` parameter: `add_address(from, n - res, _RET_IP_, to);`

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -74,18 +74,24 @@ static inline void force_uaccess_end(mm_segment_t oldfs)
 	check_object_size(to, n, false);
-	return raw_copy_from_user(to, from, n);
+	res = raw_copy_from_user(to, from, n);
+	add_address(from, n, _RET_IP_, to);
```
Please fix the length here as well.

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -127,6 +133,7 @@ _copy_from_user(void *to, const void __user *from, unsigned long n)
 	if (likely(access_ok(from, n))) {
 		instrument_copy_from_user(to, from, n);
 		res = raw_copy_from_user(to, from, n);
+		add_address(from, n, _RET_IP_, to);
```
... and here

ramosian-glider commented on kernel/entry/common.c:
```diff
@@ -121,12 +121,12 @@ noinstr void syscall_enter_from_user_mode_prepare(struct pt_regs *regs)
 static __always_inline void exit_to_user_mode(void)
-{
+{   
+        end_system_call();
```
Please fix the indentation.

ramosian-glider commented on kernel/entry/common.c:
```diff
@@ -121,12 +121,12 @@ noinstr void syscall_enter_from_user_mode_prepare(struct pt_regs *regs)
  */
 static __always_inline void exit_to_user_mode(void)
-{
+{   
```
Please remove the redundant spaces after this brace.

ramosian-glider commented on kernel/fork.c:
```diff
@@ -940,6 +940,10 @@ static struct task_struct *dup_task_struct(struct task_struct *orig, int node)
 	tsk->active_memcg = NULL;
+#endif
+#ifdef CONFIG_DF_DETECTION
+        tsk->addresses = NULL;
```
Please fix the indentation.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+#include "linux/df-detection.h"
```
Your header lives in include/linux as well as the ones below. Please insert it between them (and change quotes to angle brackets)

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+	char buf[64], buf_caller[64];
+	int size;
+	size = scnprintf(buf, sizeof(buf), "%ps", caller);
+	if (len > MAX_LEN || addr == 0 || strnstr(buf, "copy_from_user_nmi", size))
```
Can you check whether we still need copy_from_user_nmi()? Would be nice to remove it.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+		current->addresses[current->num_read].len = len;
+		current->addresses[current->num_read].caller = caller;
+		current->addresses[current->num_read].stack =
+		    df_save_stack(GFP_NOWAIT);
```
Please use tabs for indentation, unless you need to fine-align the next line with a specific character on the previous line (e.g. for function parameters).

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+
+void report(void)
+{
+	if (!check_valid_detection())
```
Please move this if-statement after all variable declarations.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+			    filter_stack(second_entries, second_nr_entries);
+			first_frame = first_entries[first_index];
+			second_frame = second_entries[second_index];
+			char first_buf[64], second_buf[64];
```
Linux kernel style doesn't allow mixing declarations and code. You can either move these two to the top of the scope, or to the declaration list at the beginning of the function.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+			if (strnstr(first_buf, "perf_copy_attr", first_len) &&
+			    strnstr(second_buf, "perf_copy_attr", second_len))
+				continue;
+			pr_err("BUG: multi-read in %ps, "
```
Please do not break user-visible strings, the possibility to search for them is more important than obeying the 80-column limit: https://www.kernel.org/doc/html/v4.10/process/coding-style.html#breaking-long-lines-and-strings

Note that it is fine to make lines longer (up to 100 characters) if that helps readability.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+			       "syscall %ps\n ",
+			       first_frame,
+			       sys_call_table[current->syscall_num]);
+			pr_err("==============================================="
```
Please don't break this string.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+		} else {
+			pr_err("BUG: multi-read in syscall %ps\n ",
+			       sys_call_table[current->syscall_num]);
+			pr_err("==============================================="
```
Same here.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+/*skipping systemcalls: (setsockopt, mount ,getsockopt ) for now because they
+ * are causing reports at boot time*/
+// TODO: moving this filter to syzkaller by enable and disable df_detection
+bool check_valid_detection(void)
```
Looking forward to seeing this function removed :)

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+	return stack_depot_save(entries, nr_entries, flags);
+}
+
+int is_intersect(struct df_address_range a, struct df_address_range b,
```
Let's make this function return bool, it is more natural here.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,275 @@
+
+void detect_intersection(void *kernel_addr)
+{
+	int i;
```
Please use a blank line to separate variable declarations from the code.

ramosian-glider commented on lib/iov_iter.c:
```diff
@@ -151,6 +152,7 @@ static int copyin(void *to, const void __user *from, size_t n)
 	if (access_ok(from, n)) {
 		instrument_copy_from_user(to, from, n);
 		n = raw_copy_from_user(to, from, n);
+		add_address(from, n, _RET_IP_, to);
```
Please fix the length argument.

ramosian-glider commented on lib/iov_iter.c:
```diff
@@ -10,6 +10,7 @@
 #include <net/checksum.h>
 #include <linux/scatterlist.h>
 #include <linux/instrumented.h>
+#include "linux/df-detection.h"
```
Please fix this #include.

ramosian-glider commented on lib/strncpy_from_user.c:
```diff
@@ -10,6 +10,7 @@
 
 #include <asm/byteorder.h>
 #include <asm/word-at-a-time.h>
+#include "linux/df-detection.h"
```
Please fix.

ramosian-glider commented on lib/strncpy_from_user.c:
```diff
@@ -120,6 +121,7 @@ long strncpy_from_user(char *dst, const char __user *src, long count)
 		if (user_read_access_begin(src, max)) {
 			retval = do_strncpy_from_user(dst, src, count, max);
 			user_read_access_end();
+			add_address(src, retval, _RET_IP_, dst);
```
do_strncopy_from_user() returns the number of bytes actually copied, so everything is correct here.

ramosian-glider commented on lib/usercopy.c:
```diff
@@ -2,6 +2,7 @@
 #include <linux/bitops.h>
 #include <linux/instrumented.h>
 #include <linux/uaccess.h>
+#include "linux/df-detection.h"
```
Please fix.

ramosian-glider commented on lib/usercopy.c:
```diff
@@ -13,6 +14,7 @@ unsigned long _copy_from_user(void *to, const void __user *from, unsigned long n
 	if (likely(access_ok(from, n))) {
 		instrument_copy_from_user(to, from, n);
 		res = raw_copy_from_user(to, from, n);
+		add_address(from, n, _RET_IP_, to);
```
Please fix the size.

Merna177 replied:
you mean sending res instead of n to add_address right?

Merna177 committed:
remove filtering the stack

Merna177 committed:
Revert

Merna177 commented:
if we remove syscall name from header of the report, are we still need it ? if not i can remove it from the argument of start_syscall function,

Merna177 committed:
Revert

Merna177 committed:
fix style

ramosian-glider commented:
> if we remove syscall name from header of the report, are we still need it ? if not i can remove it from the argument of start_syscall function,

Yes, I think it should be fine to remove it.

Merna177 committed:
changes in style and change the calling of start_system_call

Merna177 committed:
remove dump-stack

Merna177 committed:
changing in style

Merna177 committed:
revert changes

Merna177 commented:
is it good to change add_address to name more related to double fetching ?  like dfetch_add_address() or something like that.
Also is it good to change the reset of functions to static functions ?

Merna177 committed:
remove unnecessary filtering and changes in naming

ramosian-glider commented:
> is it good to change add_address to name more related to double fetching ? like dfetch_add_address() or something like that.

Yes, good idea.
Every function that is going to be called from other translation units (.c files) needs some prefix that will prevent name clashes and help understand the purpose of that function.

> Also is it good to change the reset of functions to static functions ?

If a function is only called from the same translation unit, it must be static, right.
Also, we probably want to make dfetch_add_address() reachable from the modules as well - for that you'll need to add EXPORT_SYMBOL

Merna177 committed:
changes in naming and style

Merna177 committed:
changes in styling and add some checks

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -74,18 +75,28 @@ static inline void force_uaccess_end(mm_segment_t oldfs)
 	check_object_size(to, n, false);
-	return raw_copy_from_user(to, from, n);
+	res = raw_copy_from_user(to, from, n);
+#ifdef CONFIG_DF_DETECTION
```
Kernel maintainers do not like excessive #ifdefs.
For our tools we usually do the following: declare the function in the header if the tool is enabled, and define it to an empty function if the tool is disabled. This way you can just call your function without any checks, and it will be simply optimized away if unneeded.
I.e.:
```
...
#ifdef CONFIG_DF_DETECTION
...
void dfetch_add_address(const void *addr, size_t len, unsigned long caller, void * kernel_addr);
...
#else
...
static inline void dfetch_add_address(const void *addr, size_t len, unsigned long caller, void * kernel_addr) {}
...
#endif
```
(check e.g. include/linux/kasan.h for other examples).

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -74,18 +75,28 @@ static inline void force_uaccess_end(mm_segment_t oldfs)
-	return raw_copy_from_user(to, from, n);
+	res = raw_copy_from_user(to, from, n);
+#ifdef CONFIG_DF_DETECTION
+	dfetch_add_address(from, n, _RET_IP_, to);
```
For some reason I cannot answer your comment about the size being passed to dfetch_add_address().
We need to pass the actual number of bytes that were copied.
Since copy_from_user() returns the number of bytes not copied, we need to pass `n - res` to dfetch_add_address()

Merna177 replied:
ahh, okay i will fix it

ramosian-glider commented on include/linux/uaccess.h:
```diff
@@ -74,18 +75,28 @@ static inline void force_uaccess_end(mm_segment_t oldfs)
 static __always_inline __must_check unsigned long
 __copy_from_user(void *to, const void __user *from, unsigned long n)
 {
+	unsigned long res;
```
Please leave a blank line between the variable declarations and the code.

ramosian-glider commented on kernel/fork.c:
```diff
@@ -940,6 +940,10 @@ static struct task_struct *dup_task_struct(struct task_struct *orig, int node)
 	tsk->active_memcg = NULL;
+#endif
+#ifdef CONFIG_DF_DETECTION
+        tsk->dfetch_addresses = NULL;
```
Please mind the indentation (looks like you have 8 spaces on this line and a tab on the following line)

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+		 void *kernel_addr)
+{
+	if (!in_task() || current->dfetch_addresses == NULL || current->dfetch_pairs == NULL ||
+	    addr > TASK_SIZE)
```
Nit: you could move this `addr > TASK_SIZE` check to the following if-statement, so that two `addr` checks belong together. This will also save you a line.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+		return;
+	if (current->num_read >= current->ranges_size && current->ranges_size < DFETCH_MAX_RECORDS) {
+		struct dfetch_address_range *temp =
+		    (struct dfetch_address_range *)krealloc(
```
Nit: I don't think using spaces for indentation is justified here.
Try running clang-format on this file.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+		 void *kernel_addr)
+{
+	void *a_end = (void *)((char *)a.start_address + a.len);
+	void *b_end = (void *)((char *)b.start_address + b.len);
```
Please add a blank line after this variable declaration.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,42 @@
+};
+
+void dfetch_add_address(const void *addr, size_t len, unsigned long caller, void * kernel_addr);
+void start_system_call(void);
```
I think start_system_call()/end_system_call() also need a dfetch_ prefix.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+	}
+}
+
+void report(void)
```
Please make this function static.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+	return 0;
+}
+
+void detect_intersection(void *kernel_addr)
```
Please make this function static.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+	void *b_end = (void *)((char *)b.start_address + b.len);
+	if (a.start_address <= b.start_address && a_end > b.start_address) {
+		size_t len = (char *)(a_end > b_end ? b_end : a_end) -
+			     (char *)b.start_address;
```
Please add a blank line after the declaration.
Or, even better, move the declaration to the top of the function, as you are declaring this variable in both branches anyway.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+		unsigned long diff =
+		    (char *)a.start_address - (char *)b.start_address;
+		size_t len = (char *)(a_end > b_end ? b_end : a_end) -
+			     (char *)a.start_address;
```
Please move the declaration of `len` to the top.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,209 @@
+	} else if (b.start_address <= a.start_address &&
+		   b_end > a.start_address) {
+		unsigned long diff =
+		    (char *)a.start_address - (char *)b.start_address;
```
Please insert a blank line after `diff` declaration.
Moving this declaration to the top is optional.

Merna177 committed:
changes in style and fix size in add_address

Merna177 committed:
fix spaces and declaration

Merna177 committed:
change some functions to static and fix styling of the header file

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,44 @@
+#ifndef __LINUX_DOUBLE_FETCH_DETECTION_H__
+#define __LINUX_DOUBLE_FETCH_DETECTION_H__
+
+#include <linux/types.h>
```
linux/types.h should go after linux/stackdepot.h

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,44 @@
+#define DFETCH_DISABLE          _IO('c', 255)
+#define BYTE_MAX 256
+
+/*For each address range used by copy functions*/
```
Mind the punctuation in the comments.
There should be spaces after `/*` and before `*/`. Sentences should end with a period.

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -0,0 +1,44 @@
+void dfetch_end_system_call(void){}
+void dfetch_add_address(const void *addr, size_t len, unsigned long caller, void * kernel_addr){}
+
+#endif  /*CONFIG_DF_DETECTION*/
```
Please add spaces after `/*` and before `*/` here and below.

ramosian-glider commented on include/linux/sched.h:
```diff
@@ -1315,6 +1316,16 @@ struct task_struct {
 #endif
 
+ //needed in double fetch detection buffer to store address ranges 
+#ifdef CONFIG_DF_DETECTION
```
Please declare a struct in linux/df-detection.h that will hold all these fields, so that task_struct contains just an instance of that struct instead of assorted data.

ramosian-glider commented on kernel/entry/common.c:
```diff
@@ -26,6 +27,7 @@ static __always_inline void enter_from_user_mode(struct pt_regs *regs)
 	CT_WARN_ON(ct_state() != CONTEXT_USER);
 	user_exit_irqoff();
 
+	dfetch_start_system_call();
```
If we think of user_exit_irqoff(), dfetch_start_system_call() and trace_hardirqs_off_finish() as opening brackets, and dfetch_end_system_call(), trace_hardirqs_on_prepare(), user_enter_irqoff() as closing brackets, ranges defined by these brackets should be properly nested.
I suggest moving dfetch_start_system_call to the end of enter_from_user_mode() to achieve this.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,242 @@
+	} else
+		return 0;
+	get_random_bytes(addr, len);
+	pr_err("Random injection in %px, Length %lu \n", addr, len);
```
I think it's better to move this line into get_random_bytes().
At some point the injection logic may become more complicated, and it may happen conditionally. In that case we won't need to print anything.

Merna177 replied:
you mean moving printing line into get_random_bytes()? get_random_bytes() is a function declared in random.h so in this way we will modifying an existing function

ramosian-glider replied:
Ah, sorry, I didn't realize that. Please ignore this comment.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,242 @@
+static void report(void)
+{
+	int i;
+	unsigned long *first_entries;
```
It'll be more compact if you declare first_entries and second_entries together on the same line.
Same for xxx_nr_entries.

ramosian-glider replied:
I even think replacing every pair of variables with an array will make the code more compact.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,242 @@
+}
+
+static const struct file_operations dfetch_fops = {
+    .open = nonseekable_open,
```
Please use tabs instead of spaces here.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -0,0 +1,242 @@
+{
+	switch (cmd) {
+	case DFETCH_ENABLE:
+		/* Enable DF for the current task.*/
```
You can probably move this comment in front of the function declaration and write "Enable/disable" instead of just "Enable"

Merna177 committed:
add struct double_fetch to df-detection

ramosian-glider commented on include/linux/df-detection.h:
```diff
@@ -16,19 +16,29 @@
 };
 
+struct double_fetch {
+        struct dfetch_address_range *dfetch_addresses;
```
Please fix the indentation.

ramosian-glider commented on lib/df-detection.c:
```diff
@@ -40,57 +40,66 @@ static void detect_intersection(void *kernel_addr)
-				  kernel_addr))
+	for (i = 0; i < current->dfetch.num_read; i++) {
+		if (!is_intersect(
+			current->dfetch.dfetch_addresses[i],
```
This code is using current->dfetch a lot, so it's better to declare a pointer to that struct and use it instead.
The compiler will do that anyway, but the code will also become cleaner.

Merna177 committed:
change configuration name
